extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java

```diff
@@ -477,9 +477,10 @@ public boolean satisifes(LoadingCF other) {
                     this.loading = null;
                 }
             }
+            f.complete(prj);
```

Comment:
This is tricky. I've encountered a situation when:
* the project was loading
* fired the [event at line 481](https://github.com/apache/netbeans/pull/3942/files#diff-e833b1ad0623c45c9af375ded2ba2f854c529638f4334c624969e3ea40a6e05eL481),
* it went from the listener through a series of other listeners (classpath changes), 
* then eventually attempted to obtain the GradleProject - in the same execution thread
* the future was NOT completed yet (as it would have been completed after the event dispatch ended)
* the loader was blocked
* all other gradle reloads were blocked, as the RP has throughput 1

An opinion on the fix is needed. This proposed version is simple: fire event only after the future completes. A code that does `NbGradleProject.toQuality(...).thenApply()` will progress with the loaded project, BUT then it gets a change event ... and may perform additional processing / updates, although the state is clean.
It could be possible to introduce a `ThreadLocal` that would return completed future just for the event firing thread; event will be delivered **before** the (public) Future actually completes. But this does not prevent a deadlock between more threads: firing thread waiting on X, while X is blocked on Future.get()

Reply:
This is really tricky and i'm not sure about this. I would vote that this fix is ok for now.

Reply:
Well I think this patch is reasonable. There might be better options we can come up out there, but right now it seems to be good enough. Just confirm that I'm reading it right. In this patch, we simply make the completable future completed before we are firing the notification events, so the event handlers can get the project in non-transient state?

Reply:
That was the point; the drawback now is that the client may 'over-react' since it first computes its state when the future becomes fullfilled, then receives a change event (and recomputes, hopefully to the same state).